### PR TITLE
Initial patchset for integration with debian

### PR DESCRIPTION
**Pull request by spog**
In the process of preparing a debian git-subrepo package, a set of changes had to be prepared to adapt git-subrepo to debian runtime and build system requirements. 
Changes have been grouped as follows:
  * Removed hashbangs from sourced-only files (debian policy)
  * Removed executable permission on sourced-only files (debian policy)
  * Fixed bash-completion integration with git (git sub[tab][tab] did not find subrepo)
  * Drop unnecessary subdir in usr/share/git-subrepo (only for 'make install' installation method)
  * Fixed 'make test' for debian build (debian build of package runs from a non-git git-subrepo project and for different languages)
  * Generate test git-repos instead of committed binaries (i prefer text files in a git project whenever possible)

### CODE REVIEW

**Comment by admorgan** [comment]
I am in the middle of tracking down a couple of annoying bugs. The intentions of this patch look reasonable, but it may be a bit before I dig in and determine if I can approve it. This is largely due to the fact the bugs I am tracking down are caught by our CI and therefore it makes it harder to test these changes.

**Comment by spog** [comment]
Thanks for the reply and take your time:)

**Review by admorgan — CHANGES_REQUESTED** [review]
These are all auto-generated files, see the comment "# DO NOT EDIT. This file generated by pkg/bin/generate-help-functions.pl." present in each. The source for these files needs to be updated and then run `make compgen` to regenerate the files.

**Review by admorgan — CHANGES_REQUESTED** [review]
I am sorry, but I completely stepped on this patch with the 0.4.7 release. If there is a list somewhere about what rules are expected for Debian tests I could help rework this patch.

**Review by admorgan — CHANGES_REQUESTED** [review]
I would really like to have a test for both of these methods of install, but I am not sure how to test auto-completion. Any ideas?

File: lib/git-subrepo.d/help-functions.bash
Comment (admorgan) [inline lib/git-subrepo.d/help-functions.bash:1]: This is an autogenerated file, see the comment "# DO NOT EDIT. This file generated by pkg/bin/generate-help-functions.pl." present in each. The source for these files needs to be updated and then run make compgen to regenerate the files.

File: pkg/bin/generate-completion.pl
Comment (admorgan) [inline pkg/bin/generate-completion.pl:1]: This is an autogenerated file, see the comment "# DO NOT EDIT. This file generated by pkg/bin/generate-help-functions.pl." present in each. The source for these files needs to be updated and then run make compgen to regenerate the files.

File: pkg/bin/generate-help-functions.pl
Comment (admorgan) [inline pkg/bin/generate-help-functions.pl:1]: This is an autogenerated file, see the comment "# DO NOT EDIT. This file generated by pkg/bin/generate-help-functions.pl." present in each. The source for these files needs to be updated and then run make compgen to regenerate the files.

File: share/completion.bash
Comment (admorgan) [inline share/completion.bash:1]: This is an autogenerated file, see the comment "# DO NOT EDIT. This file generated by pkg/bin/generate-help-functions.pl." present in each. The source for these files needs to be updated and then run make compgen to regenerate the files.

File: test/status.t
Comment (admorgan) [inline test/status.t:1]: I am sorry, but the 0.4.7 release changed all this (hopefully to a state you don't need this version any longer). I appologize.

File: lib/git-subrepo.d/help-functions.bash
Comment (admorgan) [reply lib/git-subrepo.d/help-functions.bash:1]: Forgot that commenting on a commit doesn't give context. Moved the comment to the file instead of the commit.

File: lib/git-subrepo [hunk 1]
```diff
@@ -12,21 +12,8 @@ set -e
...
 # Import Bash+ helper functions:
-SOURCE=${BASH_SOURCE[0]}
-while [[ -h $SOURCE ]]; do
-  DIR=$( cd -P "$( dirname "$SOURCE" )" && pwd )
-  SOURCE=$(readlink "$SOURCE")
-  [[ $SOURCE != /* ]] && SOURCE=$DIR/$SOURCE
-done
-SOURCE_DIR=$(dirname "$SOURCE")
-
-if [[ -z $GIT_SUBREPO_ROOT ]]; then
```
Comment (admorgan) [inline lib/git-subrepo:23]: GIT_SUBREPO_ROOT is advertised as a valid install method for Windows. See https://github.com/ingydotnet/git-subrepo/issues/271#issuecomment-373722235. Need to investigate if this will work with a Windows system.

**Comment by spog** [comment]
@admorgan Hi, and thanks for the review. i am currently on vacation until the end of July. I hope we can proceed then.
regards, Samo

File: Makefile
Comment (admorgan) [inline Makefile:1]: I agree that the test "source" repos should be generated by the tests instead of having their binaries in the repo. This solution doesn't work for me because the tests are frequently run outside the scope of the Makefile. Therefore this specific solution doesn't work. See the documentation and search for "prove". This logic needs to be moved to the test/setup script.

File: test/status.t
Comment (spog) [reply test/status.t:1]: No need to apologize and thanks for the info.

File: lib/git-subrepo (hunk 1 above)
Comment (spog) [reply lib/git-subrepo:23]: I imagine, compatibility with Windows is a challenge. I can not say more at this point.

File: Makefile
Comment (spog) [reply Makefile:1]: I'll try to address this according to your proposal.

**Comment by spog** [comment]
@admorgan 
> I would really like to have a test for both of these methods of install, but I am not sure how to test auto-completion. Any ideas?

I've been trying to figure out how to test auto-completion specifically for git and all i've come up is the script below, which you may call with params otherwise added to the 'git' command and the script outputs completion words in a similar manner as a real interactive bash auto-completion. The script (git-completion-test.sh) is far from perfect, but i hope it may help:
`#!/bin/bash

source /usr/share/bash-completion/bash_completion
__load_completion git

function echo_vars ()
{
	local idx=0
	echo "cur=${cur}"
	while [ -n "${words[idx]}" ]; do
		echo "words[${idx}]=${words[idx]}"
		((idx++))
	done
	echo "cword=${cword}"
	echo "prev=${prev}"
}

COMPREPLY=()
words=()
cur=
words[0]=git
cword=1
prev=git

for arg_word in $(echo "${@}"); do
	if [ -z "${arg_word}" ]; then
		break;
	fi
	words[cword]=${arg_word}
	prev=${words[((cword-1))]}
	((cword++))
done
cur=${words[(($cword-1))]}
words[(($cword-1))]=
((cword--))
prev=${words[(($cword-1))]}
#echo_vars
#echo "Check as incomplete last word: ${cur}"
__git_main

if [ -z "${COMPREPLY[0]}" ] || [ "${COMPREPLY[0]}" == "${cur} " ]; then
	words[cword]=$cur
	prev=$cur
	cur=
	((cword++))
	#echo_vars
	#echo "Re-check as complete last word: ${prev}"
	__git_main
fi

if [ -z "${COMPREPLY[0]}" ] || [ "${COMPREPLY[0]}" == "${prev} " ]; then
	#echo "No words available to complete!"
	exit 1
fi

#echo "Words available to complete:"
idx=0
while [ -n "${COMPREPLY[idx]}" ]; do
	echo -n "${COMPREPLY[idx]}"
	((idx++))
done
echo
`
best regards, Samo

File: Makefile
Comment (spog) [reply Makefile:1]: I prepared a new change by moving my Makefile changes from last two commits of this PR (regarding generation of test repos and potentially initializing the whole project as git project for cases), into the test/setup script.
Currently i have the new change only locally as modified last two commits of what constitutes this PR. But now i wander, if i force push the change into my git-subrepo fork, the original PR reference is lost. So how would you prefer to see this change presented with regard of the state of this PR?

**Comment by spog** [comment]
@admorgan, sorry for requesting a review. I am not always sure what i am clicking in github:(

**Comment by admorgan** [comment]
@spog, no worries, as is evident from my code review I have a lot of issues with their interface also. In case you have to work in github more frequently, if you do a force push to the same branch you have requested a PR on, it will automatically update the PR with the new code and even show the project what changed between the pushes.

The script is really useful and I thank you for helping with that. I am trying to expand it to ensure that "git subrepo ini" completes to the appropriate expansion and I will be adding a test for that as soon as I have some time to sort it out.

In short, Github kind of sucks as a UI; you can submit updates to this patch by force pushing on the same branch in your project and Github does something relatively sane; and I appreciate all the work you are putting into this.
